docs/_posts/2022-05-06-writing-cli-tools-with-typescript.md

```diff
@@ -0,0 +1,187 @@
+# Writing CLI tools with TypeScript
+
+## Introduction
+
+There are quite a couple of technology options when it comes to writing a *command-line interface* (CLI) tool.
+
+Most modern languages at the very least provide some kind of out-of-the-box support for command-line argument parsing. 
+
+In addtion, the ecosystems of modern languages also typically contain at least one mature and feature-rich CLI library.
+
+To give a few examples, here are some of the libraries and languages I've used in the recent couple of years to write  CLIs
+
+- [clap](https://docs.rs/clap/latest/clap/) (Rust)
+- [Click](https://docs.python-guide.org/scenarios/cli/) (Python)
+- [Picocli](https://picocli.info/picocli-2.0-groovy-scripts-on-steroids.html) (Groovy)
+- [Commander.js](https://github.com/tj/commander.js) (JavaScript)
+
+So, why would we prefer writing a CLI in TypeScript given all these other choices?
+
+Well, TypeScript is an overall good choice for a couple of reasons: 
+
+1. **compatibility** with JavaScript (TypeScript is a superset of—and by that token compatible with—JavaScript),
+2. **popularity** (TypeScript was ranked 7th in the [2021 Stack Overflow Developer Survey](https://insights.stackoverflow.com/survey/2021#section-most-popular-technologies-programming-scripting-and-markup-languages) with JavaScript being 1st),
+3. **tooling choice** (the Node.js ecosystem provides [quite a lot of CLI libraries](https://openbase.com/categories/js/best-javascript-cli-libraries)),
+4. **developer friendliness** (TypeScript is a scripted language but also has an advanced type system and static type checking which helps writing high-quality code),
+5. **performance** (since TypeScript code is, in the end, typically run by Node.js, it is in general sufficiently performant thanks to the [v8](https://v8.dev/) engine).
+
+Nonetheless, a typically non-trivial part of any TypeScript project is the set up. Writing a CLI with e.g. Python and the [Click](https://click.palletsprojects.com/en/8.1.x/) library requires little more than writing a file with the [main function](https://docs.python.org/3/library/__main__.html#packaging-considerations). The higher setup complexity of TypeScript is of course due to the nature of TypeScript being a superset of JavaScript. To run TypeScript code, you typically transpile it to JavaScript and then execute the JavaScript code with Node.js. (Yes, even [ts-node](https://github.com/TypeStrong/ts-node) does that.) So, before running a program written in TypeScript, we need to transpile it to JavaScript first. This requires tooling and therefore setup.
+
+Preferably, we also want to have a single, packaged binary as the result of building our project, rather than a script file, to reduce installation overhead on part of the user. 
+
+To assure high code quality we also typically want to include code formatting, linting, and tests.
```

Comment:
```suggestion
To assure high code quality we also typically want to include code formatting, linting and tests.
```

Reply:
I belive the comma is correct (https://www.grammarly.com/blog/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/).